RFC0003.md

```diff
@@ -0,0 +1,171 @@
+
+* PMIX\_CONNECT\_OPTIONAL: participation is optional - do not return error if procs terminate without having connected. This is the joint responsibility of the host RM (to notify if remote procs terminate early) and the PMIx server.
+
+* PMIX\_CONNECT\_EARLY\_TERM\_ALLOWED: do not generate an error if a process terminates without disconnecting. The host RM and PMIx server are responsible for adjusting completion requirements on any ongoing or future collectives for this nspace.
+
+* PMIX\_CONNECT\_XCHG\_ONLY: ensure that all participating processes have a copy of the job-level info from each nspace, but do not assign a new nspace and rank
+
+
+#### PMIx\_Connect
+Record the specified processes as "connected". Both blocking and non-blocking
+versions are provided. This means that the resource manager should treat the
+failure of any process in the specified group as a reportable event, and take
+appropriate action. Note that different resource managers may respond to
+failures in different manners.
+
+The host RM will assign a unique nspace to the resulting process group. The
+new nspace must be provided when disconnecting from the group. All procs are
+required to disconnect from the nspace prior to terminating.
+
+Note: a process can only engage in _one_ connect operation involving the identical
+set of processes at a time. However, a process _can_ be simultaneously engaged
+in multiple connect operations, each involving a different set of processes
+
+As in the case of the fence operation, the info array can be used to pass
+user-level directives regarding the algorithm to be used for the collective
+operation involved in the "connect", timeout constraints, and other options
+available from the host RM.
+
+The server is required to return any job-level info for the connecting
+processes that they might not already have - i.e., if the connect request
+involves procs from different nspaces, then each proc shall receive the
+job-level info from those nspaces other than their own.
+
+The blocking form of this call must provide a character array of size
+PMIX\_MAX\_NSLEN+1 for the assigned nspace of the resulting group, and a pointer
+to an pmix\_rank\_t location where the new rank of this process in the assigned
+nspace can be returned. Calls will return once the specified operation
+is complete (i.e., all participants have called PMIx_Connect).
+
+```
+PMIX_EXPORT pmix_status_t PMIx_Connect(const pmix_proc_t procs[], size_t nprocs,
+                                       const pmix_info_t info[], size_t ninfo,
+                                       char nspace[], pmix_rank_t *newrank);
+```
+
+The callback function for the non-blocking form of the PMIx\_Connect
+operation will be called once the operation is complete. By default, any participant
+that fails to call "connect" prior to terminating will cause the
+operation to return a "failed" status to all other participants.
```

Comment:
the uniform property (i.e. all other participants return failed) can be expensive to achieve. A less costly alternative is to relax to `any participant that fails to call "connect" prior to terminating will cause the operation to return, possibly with a "failed" status, at all other participants. However, there is a loss of functionality in that processes may connect or not non-uniformly which is hard to deal with.  

My take is that we can postpone that issue to later, if the performance gain is seizable. We can then introduce a variant with an INFO key informing that non-uniform return is ok.

Reply:
So...are you proposing a change to the language here? Or pointing out a possible future extension?

Reply:
I think the suggestion is that we go with the uniform return behavior (as written). Then, possibly, introduce a variant later (say via an info key) that relaxes the return semantics.

Reply:
Yes, there's no followup action, if we all agree on that.

Reply:
Agreed - thanks